mm2src/adex_cli/src/adex_proc/adex_proc_impl.rs

```diff
@@ -0,0 +1,182 @@
+use anyhow::{bail, Result};
+use log::{error, info, warn};
+use mm2_rpc::data::legacy::{BalanceResponse, CoinInitResponse, GetEnabledResponse, Mm2RpcResult, MmVersionResponse,
+                            OrderbookRequest, OrderbookResponse, SellBuyRequest, SellBuyResponse, Status};
+use serde_json::{json, Value as Json};
+
+use super::command::{Command, Dummy, Method};
+use super::response_handler::ResponseHandler;
+use super::OrderbookConfig;
+use crate::activation_scheme_db::get_activation_scheme;
+use crate::adex_config::AdexConfig;
+use crate::transport::Transport;
+use crate::{error_bail, warn_bail};
+
+pub(crate) struct AdexProc<'trp, 'hand, 'cfg, T: Transport, H: ResponseHandler, C: AdexConfig + ?Sized> {
+    pub transport: &'trp T,
+    pub response_handler: &'hand H,
+    pub config: &'cfg C,
+}
```

Comment:
fields shouldn't be `pub` when it's parent(struct) is `pub(crate)`

Reply:
```
error[E0451]: field `transport` of struct `AdexProc` is private
   --> src/cli.rs:110:13
    |
110 |             transport: &transport,
    |             ^^^^^^^^^^^^^^^^^^^^^ private field
```

Reply:
I guess you removing the `pub` completely instead of making it `pub(crate)`

Reply:
I thought fields can be pub if struct is pub(crate) no way to get them from anywhere no need to define them as pub(crate). Isn't it?

Reply:
To make them able to be accessed inside the crate without wordy pub(crate) everywhere. I think it's quite reasonable and not criminal trick

Reply:
So your reason is because you like `pub` word more than `pub(crate)` ? :D

Reply:
> if you could prove the opposite

1- `pub` and `pub(crate)` are provided for different purposes from the compiler.
2- these fields are not supposed to be accesed from different crates
3- that code-block literally makes no sense when you read.

If you think otherwise, you can keep it as it is then. I can't say anything extra to what I said above.

Reply:
I'm afraid this makes sense. As you can see, it must be pub, otherwise the compiler would give an error. Also, I checked that both cases get the same binary code, so here you can see for yourself:

![pub-specifier](https://github.com/KomodoPlatform/atomicDEX-API/assets/801971/f55e12f6-9341-4c9c-8031-206c8b0c92d1)

Reply:
Are you sure you are doing `pub(crate)` correctly? You shouldn't get an error. And I don't know what you are trying to achieve with assembly debugger for visibility controller

Reply:
> internal pub specifier covered with outer one.

Outer is crate-level where inner is across-crate level.

Reply:
I make sure that the code is not overloaded and is easy to read and enjoyable

Reply:
> And I don't know what you are trying to achieve with assembly debugger for visibility controller. Sure you're right it's a compile time constrains

Reply:
@ozkanonur I found an explanation, it should be enough, could I ask you to have a look please

[1](https://rust-lang.github.io/rfcs/1422-pub-restricted.html#fields-and-inherent-methods-more-public-than-self), [2](https://rust-lang.github.io/rfcs/1422-pub-restricted.html#crate-restricted-visibility)

There is written:

> "It is okay that the fields y and z and the inherent method fn callanywhere are more publicly visible than X."
> "We are allowed to restrict an inherent method, fn only_in_c, to a subtree of the module tree where X is itself visible."

Reply:
Please note in the actual RFC https://github.com/rust-lang/rfcs/blob/master/text/1422-pub-restricted.md#drawbacks it states
> However, my counter-argument to this drawback is that this feature in fact simplifies the developer's mental model. It is easier to directly encode the expected visibility of an item via pub(restriction) than to figure out the right concoction via a mix of nested mod and pub use statements. And likewise, it is easier to read it too.

so to answer your concern here:
> I make sure that the code is not overloaded and is easy to read and enjoyable
Using `pub(crate)` might not be easier to read for you but it explicitly shows the visibility of the items which is good for the developer's mental model as stated above.

Also, this RFC is about why crate visibility and similar restrictions was added and it was to avoid re-exporting items to make them accessible in a desired scope.

Reply:
> @ozkanonur I found an explanation, it should be enough, could I ask you to have a look please
> 
> [1](https://rust-lang.github.io/rfcs/1422-pub-restricted.html#fields-and-inherent-methods-more-public-than-self), [2](https://rust-lang.github.io/rfcs/1422-pub-restricted.html#crate-restricted-visibility)
> 
> There is written:
> 
> > "It is okay that the fields y and z and the inherent method fn callanywhere are more publicly visible than X."
> > "We are allowed to restrict an inherent method, fn only_in_c, to a subtree of the module tree where X is itself visible."

The reason of doing that is to make those fields accesible from other crates. In `adex-cli` which is a single crate binary program, what is the crate that will use those publicly exposed fields? There is no point to expose fields across-fields if you don't want them to be used. Those refs doesn't say anything opposite to what I been saying.

Reply:
So if you do:

```rs
pub(crate) X {
   pub y: i32
}
```

It only makes sense if a `pub fn` returns the `y` field. So devs reading that code block will assume those fields are used from `pub` functions, but in fact they are not.

Reply:
If we start abusing the design rules of Rust, we will end up having super annoying codebase which will make sense only to the writer of the code.

Reply:
If the references, examples, and answers do not make sense to you, I would recommend sticking with the principles we have been following in the mm2 tree. It's important not to apply different rules to the code we've already written, as it will make things inconsistent.

Reply:
For the record, only reason this trick is allowed is because searching algortihm for this trick was very buggy and was complicating/bloating the compiler more than expected; so it's reverted from the compiler.